### ISSUE TITLE
[clsim] do not initialize Geant4 if it is not going to use it (Trac #1756)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1748]
As you know, Geant4 can be initialized only once. This means that, if you run the example script fullSimulation.py in simprod-scripts (with a correct IceTop configuration BTW), then you get a conflict between g4-tankresponse (the default tank response) and clsim.

Normally, if you are simulating with IceTop, you are using CLSim in hybrid mode, so there is no need to use Geant4, but it is still initialized.

I am willing to help if I can get guidance in what to do. I thought it should not be a big deal to just not initialize Geant4, but it might be more complicated than I think.

Migrated from https://code.icecube.wisc.edu/ticket/1756
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:15:18", 
    "description": "As you know, Geant4 can be initialized only once. This means that, if you run the example script fullSimulation.py in simprod-scripts (with a correct IceTop configuration BTW), then you get a conflict between g4-tankresponse (the default tank response) and clsim.\n\nNormally, if you are simulating with IceTop, you are using CLSim in hybrid mode, so there is no need to use Geant4, but it is still initialized.\n\nI am willing to help if I can get guidance in what to do. I thought it should not be a big deal to just not initialize Geant4, but it might be more complicated than I think.", 
    "reporter": "jgonzalez", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067318169976", 
    "component": "combo simulation", 
    "summary": "[clsim] do not initialize Geant4 if it is not going to use it", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2016-06-23T01:10:27", 
    "milestone": "", 
    "owner": "claudio.kopper", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160623T01:15:15: david.schultz commented:

> The one easy solution is to build clsim without geant4.  I think that's `-DDISABLE_GEANT4_IN_CLSIM=True` in cmake.

**Comment by icecube-trac** [comment]
Trac update at 20160623T13:02:09: jgonzalez commented:

> Ah, I see. I can do that, but it seems a bit less than ideal to me.
> 
> Whatever one can do with ifdefs should be possible to do in C++ so one does not need to compile several versions. I could look into the DISABLE_GEANT4_IN_CLSIM ifdefs to see how to do it, unless people are explicitly against it or think it is too complicated in the long run.

**Comment by icecube-trac** [comment]
Trac update at 20160623T14:04:31: 
* olivas changed type from "defect" to "enhancement"
* olivas changed status from "new" to "assigned"
* olivas changed milestone from "" to "July 2016"
* olivas changed owner from "" to "claudio.kopper"

**Comment by icecube-trac** [comment]
Trac update at 20160721T16:46:14: 
* olivas changed status from "assigned" to "closed"
* olivas changed resolution from "" to "duplicate"

**Comment by icecube-trac** [comment]
Trac update at 20160721T18:59:18: jgonzalez commented:

> Hey Alex, you set it as duplicate, but I don't see the other one. I suppose you do not mean ticket 1391, right?

**Comment by icecube-trac** [comment]
Trac update at 20160721T19:11:19: olivas commented:

> Yeah.  At the very least they're related and maybe merged into one ticket?  It seems like the problem is the general lack of control over GEANT.  Seems like it's all-or-nothing right now.

**Comment by icecube-trac** [comment]
Trac update at 20160721T19:15:29: 
* jgonzalez changed _comment0 from:

> They are different. Bug 1391 has to do with python scripts checking for the G4 environmental variables even if Geant4 is not going to be used, while this is about G4 static global variables being initialized even if Geant4 is actually not used in CLSim (a slightly more difficult issue).
> 
> You/Claudio decide whether to merge them, but if you do merge them, then the description of the other bug should be modified to include this.

to:

> 1469128578723529


* jgonzalez commented:

> They are different. Bug 1391 has to do with python scripts checking for the G4 environment variables even if Geant4 is not going to be used, while this is about G4 static global variables being initialized even if Geant4 is actually not used in CLSim (a slightly more difficult issue).
> 
> You or Claudio decide whether to merge them, but if you do merge them, then the description of the other bug should be modified to include this.

**Comment by icecube-trac** [comment]
Trac update at 20160721T19:24:04: 
* olivas changed status from "closed" to "reopened"
* olivas changed resolution from "duplicate" to ""
* olivas commented:

> Good point.  I'll just reopen this one.

**Comment by icecube-trac** [comment]
Trac update at 20170327T19:34:20: 
* claudio.kopper commented:

> This requires some coordination with the IceTop Geant4 processing it seems.


* claudio.kopper changed milestone from "July 2016" to "December 2017"

**Comment by icecube-trac** [comment]
Trac update at 20170329T15:56:36: jgonzalez commented:

> Well, how much coordination is required depends on what we want. There are three levels of coordination:
>  1.- We make a super simulation that includes everything :-) Not going to happen.
>  2.- We write some code to switch all the Geant4 global stuff each time the modules run (switching user actions, geometry, physics lists, etc). I have done that in Auger, but this is not what I had in mind. This would allow to run CLSim and topsimulator with Geant4 in the same process. Again: this is not what I had in mind and we most probably never will do this.
>  3.- Make sure your Geant4 global variables such as the run manager are not instantiated unless CLSim is actually going to run with Geant4. This allows one to run CLSim without Geant4 + topsimulator with Geant4 and requires no coordination between topsimulator and CLSim.
> 
> This last option is rather simple once you know your code structure. People usually place G4RunManager somewhere as a class member, a global, or a static member of a class. Same with other G4 user classes (UI, physics, actions). Then it gets instantiated when the class is created or when the library is loaded. One simple solution is to hold a smart pointer instead (or sets of smart pointers), and instantiate them inside a function that gets called before doing anything Geant4-related.
>

**Comment by icecube-trac** [comment]
Trac update at 20170329T16:02:39: jgonzalez commented:

> Just as an example, g4-tankresponse puts all these things inside an "interface" class:
> http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/g4-tankresponse/trunk/private/g4-tankresponse/g4classes/G4Interface.h
> 
> This class is instantiated in the tank response service Initialize method
> http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/g4-tankresponse/trunk/private/g4-tankresponse/I3G4TankResponse.cxx#L227
> 
> So one could decide not to initialize it but do something else.

**Comment by icecube-trac** [comment]
Trac update at 20180607T16:33:26: olivas changed milestone from "December 2017" to "Autumnal Equinox 2018"

**Comment by icecube-trac** [comment]
Trac update at 20180922T10:34:39: 
* olivas commented:

> Pretty sure this is fixed on the trunk.


* olivas changed resolution from "" to "fixed"
* olivas changed status from "reopened" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:15:18: 
* olivas commented:

> Milestone Autumnal Equinox 2018 deleted


* olivas changed milestone from "Autumnal Equinox 2018" to ""
